### PR TITLE
Add pen and draw

### LINKED ISSUE #601: Add a drawPoints/draw methods to GSLayer to support FontTools pens?
Sometime one wants to work with paths in a generic way, regardless of font format, and the FontTools Pen protocol offers a nice way to abstract away from the details of TrueType, CFF, UFO, etc.
The `ttLib._TTGlyph` or defcon or ufoLib2's `Glyph` are examples.

It would be nice if GSLayer object also had something similar. 
The `glyphsLib.buider.glyph` and `.components` modules already have code to translate from GSLayer.paths and components to equivalent calls to PointPen methods. But they also contain some glyphsLib-specific code to deal with extra pieces of data that don't translate across the pen protocol.

The idea is to extract this pen-specific handling into a GSLayer.drawPoints method (and have an accompanying "draw" method that internally uses the adapter PointToSegmentPen to handle segment-oriented pens).

I drop this gist here in case anybody want to take this up:
https://gist.github.com/anthrotype/0f70ebc2143dfb01c25c47121e7e6814

### PR DESCRIPTION
**Pull request by madig**
In preparation for decomposition of non-default layer components decomposition, implement a pen and draw methods.

Closes #601.

### CODE REVIEW

File: Lib/glyphsLib/classes.py [hunk 1]
```diff
@@ -1929,6 +1937,47 @@ def applyTransform(self, transformationMatrix):
             node.position.x = x
             node.position.y = y
 
+    def draw(self, pen: AbstractPen) -> None:
+        """Draws contour into given pen."""
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:1941]:
```suggestion
        """Draws contour with the given pen."""
```

File: Lib/glyphsLib/classes.py [hunk 2]
```diff
@@ -1929,6 +1937,47 @@ def applyTransform(self, transformationMatrix):
...
+        pointPen = PointToSegmentPen(pen)
+        self.drawPoints(pointPen)
+
+    def drawPoints(self, pointPen: AbstractPointPen) -> None:
+        """Draws points of contour into given point pen."""
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:1946]:
```suggestion
        """Draws points of contour with the given point pen."""
```

File: Lib/glyphsLib/classes.py [hunk 3]
```diff
@@ -2213,6 +2262,15 @@ def bounds(self):
     #     lambda self: self.piece,
     #     lambda self, value: setattr(self, "piece", value))
 
+    def draw(self, pen: AbstractPen) -> None:
+        """Draws component with given pen."""
+        pointPen = PointToSegmentPen(pen)
+        self.drawPoints(pointPen)
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:2268]: segment pens also have a similar `addComponent` method, you can call that directly here I guess

File: Lib/glyphsLib/classes.py [hunk 4]
```diff
@@ -3188,6 +3246,28 @@ def foreground(self):
...
+        pen = SegmentToPointPen(self.getPointPen())
+        return pen
+
+    def getPointPen(self) -> AbstractPointPen:
+        """Returns a point pen for others to draw points into self."""
+        pointPen = LayerPointPen(self)
+        return pointPen
+
+    def draw(self, pen: AbstractPen) -> None:
+        """Draws glyph into given pen."""
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3260]: > 
```suggestion
        """Draws glyph with the given pen."""
```

File: Lib/glyphsLib/classes.py [hunk 5]
```diff
@@ -3188,6 +3246,28 @@ def foreground(self):
...
+        pointPen = PointToSegmentPen(pen)
+        self.drawPoints(pointPen)
+
+    def drawPoints(self, pointPen: AbstractPointPen) -> None:
+        """Draws points of glyph into given point pen."""
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3265]:
```suggestion
        """Draws points of glyph with the given point pen."""
```

File: Lib/glyphsLib/classes.py [hunk 6]
```diff
@@ -3727,3 +3807,65 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
             del self._kerning[fontMasterId][leftKey]
         if not self._kerning[fontMasterId]:
             del self._kerning[fontMasterId]
+
+
+Number = Union[int, float]
+
+
+class LayerPointPen(AbstractPointPen):
+    """A point pen.
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3816]:
```suggestion
    """A point pen to draw onto GSLayer object.
```

File: Lib/glyphsLib/classes.py [hunk 7]
```diff
@@ -3727,3 +3807,65 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
...
+        self._path: Optional[GSPath] = None
+
+    def beginPath(self, **kwargs: Any) -> None:
+        self._path = GSPath()
+
+    def endPath(self) -> None:
+        if self._path is None:
+            raise ValueError("Call beginPath first.")
+
+        self._path.closed = True
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3835]: endPath method does not a close a path. In pointPens, the open paths are marked by the first point of a contour having the "move" type.

File: Lib/glyphsLib/classes.py [hunk 8]
```diff
@@ -3727,3 +3806,77 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
             del self._kerning[fontMasterId][leftKey]
         if not self._kerning[fontMasterId]:
             del self._kerning[fontMasterId]
+
+
+Number = Union[int, float]
+
+
+class LayerPointPen(AbstractPointPen):
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3814]: how about we move this to a glyphsLib.pens module?

Comment (anthrotype) [reply Lib/glyphsLib/classes.py:3814]: nevermind, we can do that if/when we have more than one pen

File: Lib/glyphsLib/classes.py [hunk 9]
```diff
@@ -3727,3 +3806,77 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
...
+            nodetype=_to_glyphs_node_type(segmentType),
+            smooth=smooth,
+            name=name,
+        )
+        if userData:
+            node.userData = userData
+        self._path.nodes.append(node)
+
+    def addComponent(
+        self, baseGlyph: str, transformation: Transform, **kwargs: Any,
```
Comment (anthrotype) [inline Lib/glyphsLib/classes.py:3871]: transformation should be a generic 6-tuple in the pointPen protocol, you can turn that to glyphsLib's own Transform class inside the method before initialising GSComponent

**Comment by khaledhosny** [comment]
Thank you!

File: Lib/glyphsLib/classes.py [hunk 10]
```diff
@@ -3727,3 +3806,84 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
             del self._kerning[fontMasterId][leftKey]
         if not self._kerning[fontMasterId]:
             del self._kerning[fontMasterId]
+
+
+Number = Union[int, float]
```
Comment (madig) [inline Lib/glyphsLib/classes.py:3811]: Argh, need to remove this...

**Pull request merged**